main/src/ca/uwaterloo/flix/language/phase/Eraser.scala

```diff
@@ -221,13 +246,34 @@ object Eraser {
       case Int16 => Int16
       case Int32 => Int32
       case Int64 => Int64
-      case Void | AnyType | Unit | BigDecimal | BigInt | String | Regex |
-           Region | Array(_) | Lazy(_) | Tuple(_) | MonoType.Enum(_) |
-           MonoType.Struct(_, _, _) | Arrow(_, _) | RecordEmpty | RecordExtend(_, _, _) | Native(_) | Null =>
+      case Void | AnyType | Unit | BigDecimal | BigInt | String | Regex | Region | Array(_) |
+           Lazy(_) | Tuple(_) | MonoType.Enum(_, _) | MonoType.Struct(_, _) | Arrow(_, _) |
+           RecordEmpty | RecordExtend(_, _, _) | Native(_) | Null =>
         MonoType.Object
     }
   }
 
+  private def erase(tpe: Type): Type = tpe match {
+    case v@Type.Var(_, _) => v
+    case c@Type.Cst(tc, loc) => tc match {
+      case TypeConstructor.Bool => c
+      case TypeConstructor.Char => c
+      case TypeConstructor.Float32 => c
+      case TypeConstructor.Float64 => c
+      case TypeConstructor.Int8 => c
+      case TypeConstructor.Int16 => c
+      case TypeConstructor.Int32 => c
+      case TypeConstructor.Int64 => c
```

Comment:
What about TypeConstructors like: Regex, functions, records, etc? Native?

Reply:
Perhaps some of these should be explicitly mapped to Object, and the rest (which are assumed to have been erased) should throw ICE?

Reply:
The constructors you mention are still erased to object. 

I think its too much work checking the shape of the type here. It also couples the phases since the eraser really doesn't care for correctness